### DISCUSSION TITLE
Understanding issue (liquidate function in DSCEngine)

### DISCUSSION THREAD

**Discussion by komalrao1** [Cyfrin/foundry-full-course-cu#2548] [Q&A] [answered]
```solidity
   function liquidate(address collateral, address user, uint256 debtToCover)
        external
        moreThanZero(debtToCover)
        nonReentrant
    {
        uint256 startingHealthFactor = _healthFactor(user);
        if (startingHealthFactor >= MIN_HEALTH_FACTOR) {
            revert DSCEngine__HealthFactorOk();
        }
        uint256 tokenAmountFromDebtCovered = getTokenAmountFromUsd(collateral, debtToCover);
        uint256 bonusCollateral = (tokenAmountFromDebtCovered * LIQUIDATION_BONUS) / LIQUIDATION_PRECISION;
        uint256 totalCollateralToRedeem = tokenAmountFromDebtCovered + bonusCollateral;
        _redeemCollateral(user, msg.sender, collateral, totalCollateralToRedeem);
        _burnDSC(debtToCover, user, msg.sender);
        uint256 endingHealthFactor = _healthFactor(user);
        if (endingHealthFactor <= startingHealthFactor) {
            revert DSCENGINE__HealthFactorNotImproved();
        }
    }
    
   function _redeemCollateral(address from, address to, address tokenCollateralAddress, uint256 amountCollateral)
        private
    {
        s_collateralDeposited[from][tokenCollateralAddress] -= amountCollateral;
        emit CollateralRedeemed(from, to, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transfer(to, amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
    
   function _burnDSC(uint256 amountDscToBurn, address onBehalfOf, address dscFrom) private {
        s_DSCMinted[onBehalfOf] -= amountDscToBurn;
        bool success = i_dsc.transferFrom(dscFrom, address(this), amountDscToBurn);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
        i_dsc.burn(amountDscToBurn);
    }
    
```
I couldn't understand how liquidators gonna benefit by liquidating. What i understood is when some one deposit collateral and mint DSC then later when the deposits become under collateralize then other people can liquidate the positions of one who is having under collateralize . In the liquidate function it is making a call to `_burnDSC()` function and in the logic it is updating the **DSC** token balance of one who is under collateralized and then it is transferring **DSC** tokens from liquidator to **DSC** Engine then burning those tokens. 

My doubt is how liquidators gonna benefit? 

Assume a person deposited 1WETH token (price of WETH $2000) and got 1000 DSC coins. Now later WETH price decreased to $800. Now a liquidator liquidated the positions by sending 1000DSC coins to DSCEngine and burnt it. For this liquidator should get WETH tokens worth more than $1000 but since the WETH value decreased to $800 it will be a loss for liquidator right?

**Comment by EngrPips** [comment, marked as answer]
The main problem is that you don't understand how the collateralization of the protocol works. If a user deposits $2000 worth of collateral, they can only mint up to $500 worth of DSC because the protocol is overcollateralized by 200%. So in your example, if the value of their collateral drop to $800, that is $800 backing $500 worth of DSC, but because the protocol is designed to always ensure that there is double amount backing whatever is minted as DSC by any user, it will flag the user as under collateralized and then a liquidator will come and liquidate the user paying only $500 worth of DSC and getting back $550(let' assume the liquidator gets 10% extra) and there is still some amount left from the user original deposit because $800 > $550. Does this make any sense to you?

**Reply by komalrao1** [reply]
Got it. But according to our protocol it is if we deposit $2000 then we can mint $1000. 

Thank you for the explanation

**Reply by EngrPips** [reply]
If you check the `healthFactor` function carefully, you will notice that it divides the actual collateral deposited by 2 before using it to calculate the healthFactor.
